### PR TITLE
Use destroy_all to remove test users and admins

### PR DESCRIPTION
**Pull request by md5**
This lets `rake remove_test_users_and_admins` work without error even if some
of the data has already been deleted manually or in a previous run of
`rake remove_test_users_and_admins`

### CODE REVIEW

**Comment by monfresh** [comment]
Good catch, thanks! Could you please also add some tests that make sure the task can be run multiple times and that it only deletes a subset of users, and not all users?

**Comment by md5** [comment]
Do you have any examples of specs that test Rake tasks? I'm not sure off hand how I'd do that, though I suppose I could break out the code into its own simple library or add a method each to `User` and `Admin` and test those. What do you think?

**Comment by md5** [comment]
This looks like a reasonable approach possibly: http://carlosplusplus.github.io/blog/2014/02/01/testing-rake-tasks-with-rspec/

**Comment by monfresh** [comment]
Sorry, I went offline early last night. I've been using [thoughtbot's shared context](https://robots.thoughtbot.com/test-rake-tasks-like-a-boss) (referenced in the blog post you linked to). The one thing to pay attention to is that the filename, task name, and spec description title have to match. So, here's what I would do.

Define the shared context in `spec/support/shared_contexts/rake.rb`:
```ruby
require 'rake'

shared_context 'rake' do
  let(:rake)      { Rake::Application.new }
  let(:task_name) { self.class.top_level_description }
  let(:task_path) { "lib/tasks/#{task_name.split(':').first}" }
  subject         { rake[task_name] }

  def loaded_files_excluding_current_rake_file
    $LOADED_FEATURES.reject { |file| file == Rails.root.join("#{task_path}.rake").to_s }
  end

  before do
    Rake.application = rake
    Rake.application.rake_require(task_path, [Rails.root.to_s], loaded_files_excluding_current_rake_file)

    Rake::Task.define_task(:environment)
  end
end
```

Rename the task filename from `remove_users.rake` to `remove_test_users_and_admins.rake`.

Create the spec in `spec/lib/tasks/remove_test_users_and_admins_rake_spec.rb`. Here's an example that shows how to test user removal, and you can add the admin tests:

```ruby
require 'rails_helper'

describe 'remove_test_users_and_admins' do
  include_context 'rake'

  before do
    create(:user)
    create(:admin)
    allow_any_instance_of(IO).to receive(:puts)
    Rails.application.load_seed
  end

  its(:prerequisites) { should include('environment') }

  it 'only removes 2 users' do
    subject.invoke

    expect(User.count).to eq 1
  end

  it 'only removes the demo users' do
    subject.invoke 

    expect(User.pluck(:email)).to eq %w(valid@example.com)
  end

  it 'can be run multiple times' do
    subject.invoke
    subject.reenable
    subject.invoke

    expect(User.count).to eq 1
  end
end
```

**Comment by md5** [comment]
I pushed up a couple commits that to add your example (I used `--author` since I just used your code verbatim). I'm seeing this error when I run the spec:

```
     Failure/Error: Unable to find matching line from backtrace
     Bullet::Notification::UnoptimizedQueryError:
       user: mike
       N+1 Query detected
         User => [:api_applications]
         Add to your finder: :includes => [:api_applications]
       N+1 Query method call stack
         /Users/mike/Code/ohana-api/lib/tasks/remove_test_users_and_admins.rake:2:in `block in <top (required)>'
         /Users/mike/Code/ohana-api/spec/lib/tasks/remove_test_users_and_admins_rake_spec.rb:28:in `block (2 levels) in <top (required)>'
       
       /Users/mike/Code/ohana-api/lib/tasks/remove_test_users_and_admins.rake:2:in `block in <top (required)>'
       /Users/mike/Code/ohana-api/spec/lib/tasks/remove_test_users_and_admins_rake_spec.rb:28:in `block (2 levels) in <top (required)>'
     # ./spec/support/bullet.rb:8:in `block (2 levels) in <top (required)>'
```

Looks like `bullet` may not play well with `#pluck`.

**Comment by md5** [comment]
Would it make sense to add this?

```ruby
    Bullet.add_whitelist(
      type: :n_plus_one_query, class_name: 'User', association: : api_applications
    )
```

**Comment by md5** [comment]
I opened this issue to ask about support for `#pluck`: https://github.com/flyerhzm/bullet/issues/226

File: spec/lib/tasks/remove_test_users_and_admins_rake_spec.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  its(:prerequisites) { should include('environment') }
+
+  it 'only removes 2 users' do
+    subject.invoke
+
+    expect(User.count).to eq 1
+  end
+
+  it 'only removes the demo users' do
+    subject.invoke
```
Comment (md5) [inline spec/lib/tasks/remove_test_users_and_admins_rake_spec.rb]: I can fix this.

File: spec/support/shared_contexts/rake.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
+require 'rake'
+
+shared_context 'rake' do
+  let(:rake)      { Rake::Application.new }
```
Comment (md5) [inline spec/support/shared_contexts/rake.rb]: Do you actually want to follow this rule, or should the Hound config be updated to allow columnar whitespace?

**Comment by monfresh** [comment]
I think that might actually be a legitimate N+1 query since `destroy_all` calls the `dependent` callbacks to destroy the User's `api_applications`. Adding the `includes` as suggested by bullet should fix the issue:
```ruby
User.includes(:api_applications).destroy_all
```

File: spec/support/shared_contexts/rake.rb (hunk 2 above)
Comment (monfresh) [inline spec/support/shared_contexts/rake.rb]: I don't have a strong feeling about this, but I think sometimes the extra spacing makes things more readable, so I would lean towards updating the Hound config. The setting should be:
```
Style/ExtraSpacing:
  Description: Do not use unnecessary spacing.
  Enabled: false
```
If you do add this setting, please place it in alphabetical order in `.rubocop.yml`

**Comment by md5** [comment]
Still need to add the `Admin` tests here, but I wanted to push and see if this fixed Hound and Travis's failures.

File: spec/lib/tasks/remove_test_users_and_admins_rake_spec.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  it 'only removes 3 admins' do
+    subject.invoke
+
+    expect(Admin.count).to eq 1
+  end
+
+  it 'only removes the demo admins' do
+    subject.invoke
+
+    expect(Admin.pluck(:email)).to eq %w(moncef@samaritanhouse.com)
```
Comment (md5) [inline spec/lib/tasks/remove_test_users_and_admins_rake_spec.rb]: Let me know if you want me to update the `:admin` factory to use a different email address.

**Comment by monfresh** [comment]
LGTM. No need to update the admin email. It's not a real email address. I think a squash is all that's left.

**Comment by md5** [comment]
Squashed. I left the `.rubocop.yml` change in a separate commit.

**Comment by md5** [comment]
It doesn't look like Coveralls is actually going to report for the latest build. Sometimes I see a completed status on the Coveralls.io page when the notification back to GitHub still takes a couple hours to arrive, but in this case it looks like the status on the Coveralls.io page is incomplete as well.

**Comment by monfresh** [comment]
Thanks!

**Comment by md5** [comment]
:metal:

**Pull request merged**